File: LeetCode/relative_sort_array.cpp
class Solution {
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        vector<int> res;
        map<int, int> mp;
        
        for(int x : arr1) {
            mp[x]++;
        }
        
        for(int x : arr2) {
            while(mp[x]--) {
                res.push_back(x);
            }
            mp.erase(x);
        }
        
        for(auto it = mp.begin();  it != mp.end(); it++) {
            int count = it->second;
            while(count--) {
                res.push_back(it->first);
            }
        }
        
        return res;
    }
};

File: LeetCode/relative_sort_array.go
func relativeSortArray(arr1 []int, arr2 []int) []int {
    maxElem := math.MinInt

    for _, num := range arr1 {
        if num > maxElem {
            maxElem = num
        }
    }

    count := make([]int, maxElem+1)

    for _, num := range arr1 {
        count[num]++
    }

    ans := make([]int, 0, len(arr1))
    for _, num := range arr2 {
        for count[num] > 0 {
            ans = append(ans, num)
            count[num]--
        }
    }

    for i := range count {
        for count[i] > 0 {
            ans = append(ans, i)
            count[i]--
        }
    }

    return ans
}
